src/AbstractActionHandler.ts

```diff
@@ -90,36 +98,57 @@ export abstract class AbstractActionHandler {
   /**
    * Process actions against deterministically accumulating updater functions.
    */
-  protected async runUpdaters(
+  protected async applyUpdaters(
```

Comment:
This could benefit from some method extraction. It may just be the logging, but its getting complicated to read/follow

Reply:
Extracting the warnings into their own functions did help readability a lot.